internal/repository/validator.go

```diff
@@ -7,198 +7,339 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"time"
+	"log/slog"
 
 	. "github.com/cartesi/rollups-node/internal/node/model"
 	"github.com/jackc/pgx/v5"
 )
 
-const DefaultServiceTimeout = 5 * time.Minute
+var NilEpoch = errors.New("epoch cannot be nil")
 
-func (pg *Database) GetLastProcessedBlock(
+// GetOutputs returns outputs produced by inputs sent to the application
+// between start and end blocks, inclusive. Outputs are returned in ascending
+// order by index.
+func (pg *Database) GetOutputs(
 	ctx context.Context,
-	appAddress Address,
-) (uint64, error) {
-	var block uint64
-
+	application Address,
+	startBlock uint64,
+	endBlock uint64,
+) ([]*Output, error) {
 	query := `
-	SELECT
-		last_processed_block
+	SELECT 
+		o.id,
+		o.index,
+		o.raw_data,
+		o.hash,
+		o.output_hashes_siblings,
+		o.input_id
 	FROM
-		application
+		output o
+	INNER JOIN
+		input i
+	ON
+		o.input_id=i.id
 	WHERE
-		contract_address=@address`
+		i.block_number BETWEEN @startBlock AND @endBlock
+	AND
+		i.application_address=@appAddress
+	ORDER BY
+		o.index ASC
+	`
+
+	args := pgx.NamedArgs{"startBlock": startBlock, "endBlock": endBlock, "appAddress": application}
+	rows, err := pg.db.Query(ctx, query, args)
+	if err != nil {
+		return nil, fmt.Errorf("GetOutputs failed: %w", err)
+	}
+
+	var (
+		id, index, inputId   uint64
+		rawData              []byte
+		hash                 *Hash
+		outputHashesSiblings []Hash
+		outputs              []*Output
+	)
+	scans := []any{&id, &index, &rawData, &hash, &outputHashesSiblings, &inputId}
+	_, err = pgx.ForEachRow(rows, scans, func() error {
+		output := &Output{
+			Id:                   id,
+			Index:                index,
+			RawData:              rawData,
+			Hash:                 hash,
+			OutputHashesSiblings: outputHashesSiblings,
+			InputId:              inputId,
+		}
+		outputs = append(outputs, output)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("GetOutputs failed: %w", err)
+	}
+	return outputs, nil
+}
+
+// GetProcessedEpochs returns epochs from the application which had all
+// its inputs processed. Epochs are returned in ascending order by index.
+func (pg *Database) GetProcessedEpochs(ctx context.Context, application Address) ([]*Epoch, error) {
+	query := `
+	SELECT
+		id,
+    	application_address,
+        index,
+        first_block,
+		last_block,
+		claim_hash,
+		transaction_hash,
+		status
```

Comment:
Some of the lines have wrong indentation

Reply:
Something funny is happening with Github's rendering when spaces are mixed with tabs, it looks as it should in my editor. 🤷🏽‍♂️ . But I'll standardize it anyway 😉. Done in fcb5cc35fdf99dac9b5a1dd270cf21926c1083e5.